Define a Generator framework in Python

steven-johnson opened the pull request:
Work in progress, opening as draft for testing + sharing. More details + documentation to come soon.

alexreinking commented on python_bindings/src/PyGenerator.cpp:
```diff
@@ -0,0 +1,193 @@
+            });
+
+    py::object scope = m.attr("__dict__");
+    py::exec(py::str((const char *)builtin_helpers_src, builtin_helpers_src_length), scope);
```
After #6815 and the `src/` rearrangement, I don't think we'll need or want to do this binary2cpp / py::exec dance.

steven-johnson replied:
Cool, I eagerly await :-)

alexreinking commented on cmake/HalideGeneratorHelpers.cmake:
```diff
@@ -400,11 +419,12 @@ function(_Halide_add_halide_runtime RT)
                        # Defers reading the list of targets for which to generate a common runtime to CMake _generation_ time.
                        # This prevents issues where a lower GCD is required by a later Halide library linking to this runtime.
                        target=$<JOIN:$<TARGET_PROPERTY:${TARGET}.runtime,Halide_RT_TARGETS>,$<COMMA>>
                        DEPENDS "${ARG_FROM}"
+                       DEPENDS "${ARG_DEPENDS}"
```
```suggestion
                       DEPENDS "${ARG_FROM}" ${ARG_DEPENDS}
```

Does having two `DEPENDS` clauses work? That makes me nervous.

steven-johnson replied:
Oops. Maybe? Willfix.

alexreinking commented on python_bindings/src/halide/__init__.py:
```diff
@@ -22,3 +22,4 @@
 
 from halide_ import *
 from halide_ import _, _1, _2, _3, _4, _5, _6, _7, _8, _9
+from _generator_helpers import *
```
Do we need this here? This line dumps everything from `_generator_helpers` into our public API. Is that intentional? It makes sense for `halide_`, which is designed to be outward-facing.

alexreinking replied:
It should probably be `from ._generator_helpers` too (note the `.`).

steven-johnson replied:
We need some of it. E.g. `halide.GeneratorParam` which is defined in _generator_helpers but we want as part of our public API. Let me see if I can selectively import instead.

steven-johnson replied:
Using `from ._generator_helpers` didn't work for me -- stuff didn't get imported

steven-johnson replied:
After merge with latest stuff, the `from ._generator_helpers` works properly now

steven-johnson commented:
At this point, the main thing lacking here is adequate documentation (which I'm still working on); therefore I'm gonna go ahead and mark this as 'ready for review' (with the understanding that the docs will be added before landing) so that proper review can go ahead and start.

steven-johnson commented:
Lots of new stuff added to `README_python.md`, PTAL

zvookin commented on README_python.md:
```diff
@@ -1,17 +1,46 @@
 In addition to the ability to write just-in-time Halide code using Python, you
-can also use existing Halide Generators (written in C++) to produce Python
-extensions that can be used within Python code.
+can write Generators using the Python bindings, which can simplify build-system
```
Make a link from Generators to the appropriate section of this document.

steven-johnson replied:
Done

zvookin commented on README_python.md:
```diff
@@ -170,68 +201,294 @@ essentially zero cost, with storage being shared. Thus, we can usually pass it
+
+### Writing a Generator in Python
+
+A Python Generator is a class that: - has the `@hl.generator` decorator applied
```
Consider using a bullet list here for readability.

steven-johnson replied:
Ouch, it was one once -- the Markdown formatter scrambled it, fixing

zvookin commented on README_python.md:
```diff
@@ -170,68 +201,294 @@ essentially zero cost, with storage being shared. Thus, we can usually pass it
+dynamically add Inputs or Outputs to the pipeline, based on (e.g.) the values of
+`GeneratorParam` values or other external inputs
+
+Let's look at a fairly simple example:
```
The main thing missing here is a standard guideline for documentation. I'm not sure if this should be a longer comment above the class that describes all the parameters, comments on the parameter lines themselves, or a docstring in the definition of the class, but I am certain one should never write a generator without documentation on what it does and what each of its parameters, inputs, and outputs are.

steven-johnson replied:
Good point! I'll prep that as a PR-onto-this-PR, as it may need separate comment.

zvookin commented on README_python.md:
```diff
@@ -1,17 +1,46 @@
 # Halide Bindings for Python
```
We should consider changing "you" to "one" throughout this document.

steven-johnson replied:
I don't disagree, but it's not clear to me what the appropriate phrasing becomes in all such cases -- sometimes it becomes awkwardly passive-voice, whereas the 'you/your' phrasing doesn't (at least to my ear). Do you see this as simply a matter of preferred style, or are there other correctness issues I'm missing?

alexreinking replied:
I'm happy with this document using the second person. "One" sounds excessively formal to my ear. Comparable documentation from other languages uses similar second person wording, see, e.g. the Rust Book https://doc.rust-lang.org/book/ch03-04-comments.html

steven-johnson commented:
A handful of changes today, including:
-  There wasn't an implementation of `Generator::compile_to_callable()` in Python -- oops! Added it along with a test.
-  Reflecting on feedback from the last Halide meeting, I removed the public `set_generatorparam_value()` methods from Python, and instead require GeneratorParam values to be specified at Generator construction time (via an optional dict argument to the ctor). I think this will better reflect actual usage an instinct; PTAL.
-  Renamed some of the Generators under test/generator to ensure there is clarity as to C++ vs Python in the name of the Generator (when the 'same' one is implemented both ways)

abadams commented on src/Pipeline.cpp:
```diff
@@ -665,7 +674,7 @@ Callable Pipeline::compile_to_callable(const std::vector<Argument> &args_in, con
     for (const auto &out : outputs) {
         for (Type t : out.output_types()) {
-            args.emplace_back(out.name(), Argument::OutputBuffer, t, out.dimensions(), ArgumentEstimates{});
+            args.emplace_back(strip_uniquifier(out.name()), Argument::OutputBuffer, t, out.dimensions(), ArgumentEstimates{});
```
What does this stripped name get used for? It's currently legal to write nonsense like:

```
#include "Halide.h"

using namespace Halide;

int main(int argc, char **argv) {
    Func f1("f"), f2("f");
    Var x;
    f1(x) = x;
    f2(x) = x + 1;

    Pipeline p({f1, f2});

    p.compile_jit();

    return 0;
}
```

but that still seems to work in this branch

steven-johnson replied:
OK, revisiting this: the issue here is that we are building up a list of Arguments to use for the Callable; for most uses, the name is just "decoration" (e.g. for annotating the function prototype in the header). In this case, though, we're going to use them to populate the list of `kwargs` in the Python wrappers, so that we can do something like `simple(buffer_input=b_in, float_arg=float_in, simple_output=b_out)`. 

The gotcha, of course, is that Outputs in a Generator are essentially Funcs, and the names get uniquified as a result, so if you the same Generator multiple times (or different Generators with matching output names), you'll get outputs with names like `simple_output$4`, which are unusable as kwargs.

This change was really just a quick hack, on the assumption that any Argument we see that ends in `$NUM` is just a uniquified name and that we can trim off that bit to get the 'real' name.

I can revisit the code paths and see if there is way to ensure that the 'original' Output names get transmitted from the Generator all the way down this path, but it would likely take the addition of some new data channel to make that happen reliably, which is why I took the quick way out.

Given the use case here, I suspect that just doing the stripping is safe, with the caveat that you *could* construct a pipeline like this that would be ambiguous to call via kwargs in Python -- but not via a Generator, since they enforce that all inputs and outputs have unique names -- so it would be a case of "don't do that".

steven-johnson replied:
(Gentle ping to @abadams for feedback on my feedback)

abadams replied:
I'm mostly concerned about the non-python non-Generator case, because as you say name collisions are checked in a Generator. I'm worried that we'll have two sets of out-of-sync names floating around that produce hard-to-debug errors later due to the mismatch if someone does something creative with a Callable, e.g. by using names from Callable::arguments() to go look up into a map of Funcs by name. 

Maybe it would be safer to strip the $* at the point where the kwargs are populated.

steven-johnson replied:
Good call. f613271835688244830b68379de2543fb9bfc59b has the fix, I think. PTAL.

alexreinking commented on README_python.md:
```diff
@@ -1,17 +1,47 @@
+<!-- /MarkdownTOC -->
+
 Halide provides Python bindings for most of its public API. Only Python 3.x is
 supported; at this time, Python 3.8 (or higher) is recommended. The Python
```
We should reword this into a requirement, I think.

alexreinking commented on README_python.md:
```diff
@@ -1,17 +1,47 @@
 
-(Note that the ability to write Halide Generators in Python is *not* supported
-at present, but is expected to be supported soon.)
+You can also use existing Halide Generators (written in C++) to produce Python
```
> (written in C++)

but not the new ones in Python?

steven-johnson replied:
Fixed.

alexreinking commented on README_python.md:
````diff
@@ -62,7 +92,7 @@ with some differences where the C++ idiom is either inappropriate or impossible:
     Buffer<>(Type t, vector<int> extents, string name = "");
     ```
 
-    in Python, only the second variant is provided.
````
Above, is written "variadic argumentlist" (missing space). GitHub's interface doesn't let me comment too far away from a change, ugh.

alexreinking replied:
Some of the entries below might need revision. For instance, "No mechanism is provided for overriding any runtime functions from Python." probably doesn't apply to the Python generators.

steven-johnson replied:
Updated to clarify that this applies to JIT code, and added a section on the state of runtime functions for AOT code.

alexreinking commented on README_python.md:
```diff
@@ -170,35 +200,240 @@ essentially zero cost, with storage being shared. Thus, we can usually pass it
-## Using Halide Generators from Python
+## Halide Generators In Python
+
+In Halide, a "Generator" is a unit of encapsulation for Halide code. It is
```
It is **a**

steven-johnson replied:
Done

alexreinking commented on README_python.md:
```diff
@@ -170,35 +200,240 @@ essentially zero cost, with storage being shared. Thus, we can usually pass it
+## Halide Generators In Python
+
+In Halide, a "Generator" is a unit of encapsulation for Halide code. It is
+self-contained piece of code that can: - Produce a chunk of Halide IR (in the
```
Should this be a real list?

steven-johnson replied:
jfc, the markdown formatter I use keeps un-formatting this from a real list, fixed

alexreinking commented on README_python.md:
```diff
@@ -289,6 +552,201 @@ But if you construct the numpy arrays yourself (like above), you can pass
+
+    output = hl.OutputBuffer(hl.UInt(8), 2)
+
+    def configure(g):
```
Do we want to use the `g = self` idiom here? Maybe I missed it, but if we want to advocate for using `g` in place of `self` for that argument name, we should call it out and indicate that it's just for terseness and isn't magic.

steven-johnson replied:
Yep, will call this out earlier and more emphatically

steven-johnson replied:
Edit: after looking at this some I wonder if this is a mistake in our style thoughts -- if you didn't notice the docs on a read-through, other folks won't either, and will wonder WTF is going on. Maybe it would be clearer to have the convention (in our code at least) be

```
    def generate(self):
        g = self  # for terseness in expressions
       ...
```

? 

(I won't make any changes here till I hear your thoughts)

steven-johnson replied:
Any thoughts on this one?

alexreinking commented on README_python.md:
```diff
@@ -170,35 +200,240 @@ essentially zero cost, with storage being shared. Thus, we can usually pass it
+
+It is required that the `generate()` method be defined by the Generator.
+
+(Note that, by convention, Halide Generators use `g` instead of `self` in their
+`generate()` method to make the expression language terser; this is not in any
+way required, but is recommended.)
```
Whoops, I missed this earlier. This needs to be called out more; somewhere higher up than four heading levels deep.

steven-johnson replied:
See above

alexreinking commented on README_python.md:
```diff
@@ -170,35 +200,240 @@ essentially zero cost, with storage being shared. Thus, we can usually pass it
-                   GENERATOR my_filter_generator
-                   SOURCES my_filter_generator.cpp
+add_halide_library(xor_filter
+                   GENERATOR logical_op_generator
+                   SOURCES logical_op_generator.py
```
I'm confused by this... shouldn't this just be `xor_filter FROM logical_op_generator.py`?

steven-johnson replied:
Yep, fixed

alexreinking commented on README_python.md:
```diff
@@ -289,6 +552,201 @@ But if you construct the numpy arrays yourself (like above), you can pass
+    arguably just as easy, if not easier.
+-   `get_externs_map()`: this allows registering ExternalCode objects to be
+    appended to the Generator's code. In our experience, this feature is very
+    rarely used. We consider adding this in the future if necessary.
```
We will

alexreinking reviewed (changes requested):
I still have to look at the main event, `_generator_helpers.py`, but there's plenty to act on now.

steven-johnson commented on cmake/HalideGeneratorHelpers.cmake:
```diff
@@ -68,26 +72,47 @@ function(add_halide_generator TARGET)
+            add_custom_target(${TARGET} ALL)
+            set_property(TARGET ${TARGET} PROPERTY Halide_PYTHON_GENERATOR_SOURCE "${CMAKE_CURRENT_SOURCE_DIR}/${ARG_SOURCES}")
+
+            # TODO: what do we need to do for PACKAGE_NAME PACKAGE_NAMESPACE EXPORT_FILE in this case?
```
Let be more clear: I'm not entirely clear on what these are for when using C++ Generators. So I definitely don't know how to set up for python. Leaving this one open for now, I guess.

alexreinking replied:
Let's not worry about this for _this_ PR, but please create an issue assigned to the two of us along the lines of "Make Python generators work in cross compiling scenarios"

steven-johnson replied:
https://github.com/halide/Halide/issues/7014

steven-johnson commented on python_bindings/src/halide/halide_/PyGenerator.cpp:
```diff
@@ -0,0 +1,200 @@
+
+using Halide::Internal::AbstractGenerator;
+using Halide::Internal::AbstractGeneratorPtr;
+#ifdef HALIDE_ALLOW_GENERATOR_EXTERNAL_CODE
```
It's not; see https://github.com/halide/Halide/pull/6844

steven-johnson commented on python_bindings/test/generators/complexpy_generator.py:
```diff
@@ -0,0 +1,66 @@
+
+@hl.generator(name = "complexpy")
+class ComplexPy:
+    vectorize = hl.GeneratorParam(True)
```
Apples and oranges; this call expects a plain Python object. But I see your point.

steven-johnson commented on python_bindings/src/halide/halide_/PyType.h:
```diff
@@ -3,6 +3,11 @@
 
 #include "PyHalide.h"
 
+// This is an expedient: a unique Handle type that we use as a placeholder
+// in Python generators. It should never be visible to the end user.
+struct NoneType {};
+HALIDE_DECLARE_EXTERN_STRUCT_TYPE(NoneType);
```
Never visible to the user, but yeah, possible confusing; I'll rename.

steven-johnson commented on cmake/HalideGeneratorHelpers.cmake:
```diff
@@ -200,17 +224,25 @@ function(add_halide_library TARGET)
+    get_property(py_src TARGET ${ARG_FROM} PROPERTY Halide_PYTHON_GENERATOR_SOURCE)
+    if (py_src)
+        set(PYTHONPATH "$<TARGET_FILE_DIR:Halide::Python>/..")
+        set(GENERATOR_CMD ${CMAKE_COMMAND} -E env PYTHONPATH=${PYTHONPATH} ${Python3_EXECUTABLE} $<SHELL_PATH:${py_src}>)
```
(1) How could it be anything else?
(2) How would we detect and/or ensure?

steven-johnson commented on cmake/HalideGeneratorHelpers.cmake:
```diff
@@ -100,11 +125,10 @@ function(add_halide_generator TARGET)
+        file(WRITE "${generated_cpp}" "${stub_text}")
 
-        Python3_add_library(${TARGET}_pystub MODULE WITH_SOABI "${CMAKE_CURRENT_BINARY_DIR}/${TARGET}.py_stub_generated.cpp" ${ARG_SOURCES})
+        Python3_add_library(${TARGET}_pystub MODULE WITH_SOABI "${generated_cpp}" ${ARG_SOURCES})
```
Hm, if building WITH_PYTHON_BINDINGS=OFF then yes, I suppose. But this PR builds just fine that way (several of the buildbots only test that way), probably because we never attempt to use any of the python-related stuff when this is disabled. Not sure if we should wrap this (and other) sections in WITH_PYTHON_BINDINGS checks, vs just let people's code fail to build when they try to use Python with a non-Python Halide.

steven-johnson replied:
Actually, yeah, it's definitely possible for out-of-tree builds (eg apps): if they don't explicitly call it, it won't have been called. So the question is: should we do so here lazily? Or (alternately) do something like

`if (NOT TARGET Halide::Python) message(FATAL_ERROR "You must call find_package(Python3) in order to use this rule with Python Generators"`

steven-johnson replied:
Resolved (I think) by b789450926c2ca64bad5700bed4faae11be8baa3

steven-johnson commented on python_bindings/test/generators/complexpy_generator.py:
```diff
@@ -0,0 +1,66 @@
+    vectorize = hl.GeneratorParam(True)
+    extra_input_name = hl.GeneratorParam("")
+
+    typed_buffer_input = hl.InputBuffer(hl.UInt(8), 3)
```
I think I'm going to try out this (and/or the native-python-type allowances) as a PR-to-this-PR, so we can see what it feels like as an additive change.

steven-johnson commented on python_bindings/test/generators/bitpy_generator.py:
```diff
@@ -0,0 +1,19 @@
+
+@hl.generator(name = "bitpy")
+class BitGenerator:
+    bit_input = hl.InputBuffer(hl.Bool(), 1)
```
I think I'm going to try out this as a PR-to-this-PR, so we can see what it feels like as an additive change.

alexreinking reviewed (changes requested):
Still reviewing `_generator_helpers.py`, just publishing a couple comments now.

alexreinking commented on python_bindings/src/halide/_generator_helpers.py:
```diff
@@ -0,0 +1,806 @@
+            _check(not self.autoscheduler().name, "The GeneratorParam %s cannot be set more than once" % name)
+            self.autoscheduler().name = value
+        elif name.startswith("autoscheduler."):
+            sub_key = name[14:]
```
magic number here...

steven-johnson replied:
Done

alexreinking commented on python_bindings/src/halide/_generator_helpers.py:
```diff
@@ -0,0 +1,806 @@
+def generator(name=""):
+    # This code relies on dicts preserving key-insertion order, which is only
+    # guaranteed for all Python implementations as of v3.7.
+    assert sys.version_info >= (3, 7), "Halide Generators require Python 3.7 or later."
```
Aha. this should not be an `assert`, but a `_check`.

alexreinking reviewed (changes requested):
Mostly quibbly stuff. Ping when when resolved and I'll be quick about re-reviewing.

steven-johnson commented on python_bindings/src/halide/_generator_helpers.py:
```diff
@@ -0,0 +1,806 @@
+# yapf --style="{based_on_style: google, column_limit: 120, indent_width: 4}"
+
+
+# 'print()' is consumed by `halide.print()` which is implicitly present;
```
It's part of the Halide API. `Expr print(Expr, ...);`

steven-johnson replied:
(So, no, it can't really be renamed.)

steven-johnson commented:
Opened https://github.com/halide/Halide/issues/7014

steven-johnson commented on python_bindings/src/halide/_generator_helpers.py:
```diff
@@ -0,0 +1,806 @@
+
+# 'print()' is consumed by `halide.print()` which is implicitly present;
+# here's our quick-n-dirty wrapper for debugging:
+def _print(*args):
+
+    def write(data):
+        sys.stdout.write(str(data))
+
+    for i, arg in enumerate(args):
+        if i:
+            write(" ")
+        write(arg)
+    write('\n')
```
I left it in because I kept needing to re-add it for debugging. I'll try the builtins.print approach and if that work I'll use that instead.

steven-johnson commented on python_bindings/src/halide/_generator_helpers.py:
```diff
@@ -0,0 +1,806 @@
+    s = str(name)
+    if "__" in s:
+        return False
+    if not s[0].isalpha():
+        return False
+    for c in s[1:]:
+        if not (c.isalnum() or c == "_"):
+            return False
+    return True
```
Nah, that makes it even more permissive. The goal here is to *exactly* match the rules that C++ Generators use. I'll revise to ensure it doesn't include other Unicode areas (yet).

steven-johnson commented on python_bindings/src/halide/_generator_helpers.py:
```diff
@@ -0,0 +1,806 @@
+# also add some syntactic sugar, to allow users to alias
+# bool -> UInt(1), int -> Int(32), float -> Float(32)
+def _sanitize_type(t: object) -> Type:
+    if t is None:
```
Done

steven-johnson commented on python_bindings/src/halide/_generator_helpers.py:
```diff
@@ -0,0 +1,806 @@
+    # Always treat _UnspecifiedType as a non-type
+    if types is None:
+        types = []
+    elif isinstance(types, Type) and types == _UnspecifiedType():
```
Because `_UnspecifiedType()` returns a `hl.Type`, not a Python `Type`

steven-johnson commented:
PTAL

alexreinking reviewed (approved):
This LGTM now, though I understand there are some `pytype` issues?

steven-johnson commented:
> This LGTM now, though I understand there are some `pytype` issues?

Yeah... the issue is that pytype is the preferred type checker at Google, and it's a *static* type checker. Currently we use the `@generator` decorator to inject `hl.Generator` as a base class, but there is (apparently) no way to indicate this to pytype, with the result being that the type checker doesn't know that it's a subclass, and complains when any inherited methods are called (e.g. `natural_vector_size()`). 

This could be fixed by requiring that user Generators explicitly descend from `hl.Generator` (and having the decorator check that is the case), which is disappointing, as it adds unnecessary redundancy in the Generator declarations, eg..

```
@hl.generator()
class MyGen(hl.Generator):
```

For the sake of argument, we *could* have it both ways: have the decorator verify that it is a subclass, and inject the baseclass only if it isn't, but that has a subtle downside: people coding without pytype would likely omit the explicit base class, but then they'd produce code that would need to be modified when getting imported and used in a pytype environment. This would be a huge nuisance in the long run, with the (likely) result that people end up adding the explicit inheritance always just to dodge complaints.

At this point I'm inclined to just require the explicit inheritance in order to get around this problem, but I'm not happy about it (I find it especially unsatisfying that apparently PyType doesn't have a way to annotate dynamically-generated classes, because that's A Thing That Happens in dynamic languages like Python).
